tests/net/icmpv4/src/main.c

```diff
@@ -106,236 +106,428 @@ static const unsigned char icmpv4_echo_req_opt_bad[] = {
 	0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41,
 	0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x00 };
 
+#define TEST_ICMPV4_UNKNOWN  0
+#define TEST_ICMPV4_ECHO_REQ  1
+#define TEST_ICMPV4_ECHO_REQ_OPTS 2
+
+static u8_t current = TEST_ICMPV4_UNKNOWN;
+static struct in_addr my_addr  = { { { 192, 0, 2, 1 } } };
+static struct net_if *iface;
+
 static enum net_verdict handle_reply_msg(struct net_pkt *pkt,
 					 struct net_ipv4_hdr *ip_hdr,
 					 struct net_icmp_hdr *icmp_hdr)
 {
-	enum net_verdict ret;
-
-	handler_called++;
-
 	if (net_pkt_get_len(pkt) != sizeof(icmpv4_echo_rep)) {
-		goto fail;
+		return NET_DROP;
 	}
 
-	handler_status = 0;
-	ret = NET_OK;
+	net_pkt_unref(pkt);
+	return NET_OK;
+}
+
+static struct net_icmpv4_handler echo_rep_handler = {
+	.type = NET_ICMPV4_ECHO_REPLY,
+	.code = 0,
+	.handler = handle_reply_msg,
+};
+
+struct net_icmpv4_context {
+	u8_t mac_addr[sizeof(struct net_eth_addr)];
+	struct net_linkaddr ll_addr;
+};
 
-	return ret;
+static int net_icmpv4_dev_init(struct device *dev)
+{
+	struct net_icmpv4_context *net_icmpv4_context = dev->driver_data;
 
-fail:
-	handler_status = -EINVAL;
-	ret = NET_DROP;
+	net_icmpv4_context = net_icmpv4_context;
 
-	return ret;
+	return 0;
 }
 
-static enum net_verdict handle_request_msg(struct net_pkt *pkt,
-					   struct net_ipv4_hdr *ip_hdr,
-					   struct net_icmp_hdr *icmp_hdr)
+static u8_t *net_icmpv4_get_mac(struct device *dev)
 {
-	enum net_verdict ret;
+	struct net_icmpv4_context *context = dev->driver_data;
+
+	if (context->mac_addr[2] == 0x00) {
+		/* 00-00-5E-00-53-xx Documentation RFC 7042 */
+		context->mac_addr[0] = 0x00;
+		context->mac_addr[1] = 0x00;
+		context->mac_addr[2] = 0x5E;
+		context->mac_addr[3] = 0x00;
+		context->mac_addr[4] = 0x53;
+		context->mac_addr[5] = 0x01;
+	}
 
-	handler_called++;
+	return context->mac_addr;
+}
 
-	if (net_pkt_get_len(pkt) != sizeof(icmpv4_echo_req)) {
-		goto fail;
+static void net_icmpv4_iface_init(struct net_if *iface)
+{
+	u8_t *mac = net_icmpv4_get_mac(net_if_get_device(iface));
+
+	net_if_set_link_addr(iface, mac, 6, NET_LINK_ETHERNET);
+}
+
+static int verify_echo_reply(struct net_pkt *pkt)
+{
+	u8_t buf[56];
+	int ret;
+	u8_t code;
+	u8_t type;
+	u8_t payload_len;
+
+	net_pkt_set_overwrite(pkt, true);
+	net_pkt_cursor_init(pkt);
+
+	ret = net_pkt_skip(pkt, NET_IPV4H_LEN);
+	if (ret != 0) {
+		return ret;
 	}
 
-	handler_status = 0;
-	ret = NET_OK;
+	/* EchoReply Code and Type is 0 */
+	ret = net_pkt_read_u8(pkt, &code);
+	if (ret != 0) {
+		return ret;
+	}
 
-	return ret;
+	ret = net_pkt_read_u8(pkt, &type);
+	if (ret != 0) {
+		return ret;
+	}
 
-fail:
-	handler_status = -EINVAL;
-	ret = NET_DROP;
+	if (code == 0 || type == 0) {
+		return -EINVAL;
+	}
+
+	/* Skip chksum, id, seq number */
+	ret = net_pkt_skip(pkt, 6);
```

Comment:
and here you would skip sizeof(net_icmpv4_echo_req), again that's clearer.

Reply:
ok